### PR TITLE
Fixed issue [security] #19869: All group deletion accessible by any admin with one survey access

### PR DESCRIPTION
**Pull request by Shnoulle**

### CODE REVIEW

File: application/models/QuestionGroup.php [hunk 1]
```diff
@@ -163,14 +163,16 @@ public function insertRecords($data)
...
      * Returns affected rows of question group table (should be 1 or null)
      * @param integer $groupId
-     * @param integer $surveyId
+     * @param integer|null $surveyId deprecated
      * @return int|null
      */
-    public static function deleteWithDependency($groupId, $surveyId)
+    public static function deleteWithDependency($groupId, $surveyId = null)
     {
+        $QuestionGroup ::self::model()->findByPk($groupId);
```
Comment (gabrieljenik) [inline application/models/QuestionGroup.php:171]: What if the QueestionGroup is not found? Shouldn't we have a small check after query?

Comment (Shnoulle) [reply application/models/QuestionGroup.php:171]: Throw 500 error maybe ?

File: application/models/QuestionGroup.php [hunk 2]
```diff
@@ -163,14 +163,16 @@ public function insertRecords($data)
...
         // Abort if the survey is active
-        $surveyIsActive = Survey::model()->findByPk($surveyId)->active !== 'N';
+        $surveyIsActive = Survey::model()->findByPk($QuestionGroup->sid)->active !== 'N';
         if ($surveyIsActive) {
+            /* TODO : replace by a cleaner way to send error : model muts not use HTML only error system */
             Yii::app()->user->setFlash('error', gT("Can't delete question group when the survey is active"));
```
Comment (Shnoulle) [inline application/models/QuestionGroup.php:176]: @gabrieljenik do you think it's better to send a 500 error here ? Check of active survey must be done by controller in my opinion. Then if it(s not done : it' a 500

Comment (Shnoulle) [reply application/models/QuestionGroup.php:176]: It's done in LimeSurvey\Models\Services\QuestionGroupService->checkDeletePermission i throw a 500 here

File: application/models/QuestionGroup.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/QuestionGroup.php]: Thorw 500 (but must be a 404 : i report the issue)

**Comment by Shnoulle** [comment]
[19880: PermissionDeniedException and NotFoundException send 500 error](https://bugs.limesurvey.org/view.php?id=19880)

File: application/models/QuestionGroup.php [hunk 3]
```diff
@@ -163,16 +168,20 @@ public function insertRecords($data)
...
+        if (empty($QuestionGroup)) {
+            throw new NotFoundException(gt('Group not found'));
+        }
         // Abort if the survey is active
-        $surveyIsActive = Survey::model()->findByPk($surveyId)->active !== 'N';
+        $surveyIsActive = Survey::model()->findByPk($QuestionGroup->sid)->active !== 'N';
         if ($surveyIsActive) {
-            Yii::app()->user->setFlash('error', gT("Can't delete question group when the survey is active"));
-            return null;
+            throw new PermissionDeniedException(gt("Can't delete question group when the survey is active"));
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/models/QuestionGroup.php:184]: Is this the right kind of exception to be shown? Maybe BadRequestException.php is better fit?

[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/QuestionGroup.php:184]: 1. All LimeSurvey exception throw 500 error :)
2. GUI send a PermissionDeniedException https://github.com/LimeSurvey/LimeSurvey/blob/f10805aee259d9cfba72c35e3c5e6833a714519d/application/models/services/QuestionGroupService.php#L572

[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/QuestionGroup.php:184]: But you're right

**Review by gabrieljenik — APPROVED** [review]
Code structure looks good.

**Pull request merged**